org.lflang/src/org/lflang/generator/ts/TSGenerator.kt

```diff
@@ -245,6 +242,48 @@ class TSGenerator(
         }
     }
 
+
+    /**
+     * If a main or federted reactor has been declared, create a ReactorInstance
+     * for this top level. This will also assign levels to reactions, then,
+     * if the program is federated, perform an AST transformation to disconnect
+     * connections between federates.
+     */
+    private fun createMainReactorInstance() {
+        if (mainDef != null) {
+            if (main == null) {
+                // Recursively build instances. This is done once because
+                // it is the same for all federates.
+                main = ReactorInstance(
+                    ASTUtils.toDefinition(mainDef.reactorClass), errorReporter,
+                    unorderedReactions
+                )
+                val reactionInstanceGraph = main.assignLevels()
```

Comment:
Does this level assignment do anything here for reactor-ts?

Reply:
Not right now. I don't think we really use reactor instance in reactor-ts.

I think I just tried to do the same as the CGenerator.

Would you have any concern or advice?

Reply:
The old FIXME still applies here. For the purposes of correct level assignment (or correct topological sorting), we add phantom connections from source federates to destination federates that could cause compile errors or runtime errors if the source federate tries to write on those phantom connections. `removeRemoteFederateConnectionPorts(main)` just removes these phantom connections to avoid errors. This removal could lead to an incorrect topological sort in reactor-ts. We have a better potential solution for this, I think, which is to generate a [causality interface](https://github.com/lf-lang/lingua-franca/pull/1221#issue-1268041718) for the source federate in the destination federate instead of creating phantom connections.

Reply:
> The old FIXME still applies here.

Sorry, I don't think I have any idea about the old FIXME you're referring to...

> removeRemoteFederateConnectionPorts(main) just removes these phantom connections to avoid errors. This removal could lead to an incorrect topological sort in reactor-ts. We have a better potential solution for this, I think, which is to generate a https://github.com/lf-lang/lingua-franca/pull/1221#issue-1268041718 for the source federate in the destination federate instead of creating phantom connections.

I see. Thank you for your suggestion! But I'm not sure if we want to solve this problem here altogether. Would it be ok if I leave a TODO or FIXME here? Do you think getting away with `removeRemoteFederateConnectionPorts(main)` for now is the right approach for the moment?

Reply:
> Do you think getting away with removeRemoteFederateConnectionPorts(main) for now is the right approach for the moment?

I think so. The causality interface is set to be part of the FedGenerator effort anyway.

Reply:
> Sorry, I don't think I have any idea about the old FIXME you're referring to...

Oh, sorry, I was referring to [this](https://github.com/lf-lang/lingua-franca/blob/2011ccacb472451651d7e0d391127c8060fc9e54/org.lflang/src/org/lflang/generator/ts/TSGenerator.kt#L152).

Reply:
I tried deleting `removeRemoteFederateConnectionPorts(main)`, and I got an error like this.

```
lfc: error: Mixed triggers and effects from different federates. This is not permitted
 --> test/TypeScript/src/federated/DistributedLoopedPhysicalAction.lf:70:1
   |
69 | 
   | >>>>>>>>>>>>>>
70 | federated reactor DistributedLoopedPhysicalAction {
71 |     sender = new Sender();
   | ...
74 |     sender.out -> receiver.inp;
75 | }
   | < Mixed triggers and effects from different federates. This is not permitted
76 | 
```

Could we keep this for now? Would you have any suggestions?

Reply:
Oh, sorry, I misquoted your comment earlier. I meant to agree with the following:

> Would it be ok if I leave a TODO or FIXME here? 

I think we should keep `removeRemoteFederateConnectionPorts` for now, but leave a FIXME.

Reply:
Thanks @Soroosh129 ! I left a FIXME.